src/handler/core.rs

```diff
@@ -508,4 +828,192 @@ mod tests {
         // Chained non-overlapping filters do not allow anything.
         assert(filter_a().chain(filter_b()).endpoint(|| async {}), hashset! {});
     }
+
+    #[tokio::test]
+    async fn test_type_check() {
+        #[derive(Clone)]
+        struct A;
+        #[derive(Clone)]
+        struct B;
+        #[derive(Clone)]
+        struct C;
+
+        let deps = deps![A, B, C];
+
+        // Type checking an entry must always succeed.
+        type_check(&HandlerSignature::Entry, &deps);
+
+        // The case for the same handler's input types and those in the container.
+        let input_types = vec![Type::of::<A>(), Type::of::<B>(), Type::of::<C>()];
+        type_check(
+            &HandlerSignature::Other {
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
+                output_types: hashset! {},
+                obligations: hashmap! { /* Must not be used. */ },
+            },
+            &deps,
+        );
+
+        // The case for handler's input types and a superset thereof in the container.
+        let input_types = vec![Type::of::<A>(), Type::of::<B>()];
+        type_check(
+            &HandlerSignature::Other {
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
+                output_types: hashset! {},
+                obligations: hashmap! { /* Must not be used. */ },
+            },
+            &deps,
+        );
+    }
+
+    #[test]
+    // Cannot specify the exact panic message because `HashSet` is iterated in a random order.
+    #[should_panic]
+    fn type_check_panic() {
+        #[derive(Clone)]
+        struct A;
+        #[derive(Clone)]
+        struct B;
+        #[derive(Clone)]
+        struct C;
+
+        let input_types = vec![Type::of::<A>(), Type::of::<B>(), Type::of::<C>()];
+
+        type_check(
+            &HandlerSignature::Other {
+                input_types: HashSet::from_iter(input_types.iter().cloned()),
+                output_types: hashset! {},
+                obligations: hashmap! { /* Must not be used. */ },
```

Comment:
What does `/* Must not be used */` means?

Reply:
`obligations` are accessed by `type_check` only for printing purposes, when a typing error occurs. Since there is no error here, we can just leave `obligations` empty.